File: Server_Code/Packet/BinaryWriter.cpp
#include "BinaryWriter.h"

void BinaryWriter::WriteUInt8(uint8_t value)
{
    buffer.push_back(static_cast<char>(value));
}

void BinaryWriter::WriteUInt16(uint16_t value) {
    char bytes[2];
    std::memcpy(bytes, &value, sizeof(uint16_t));
    buffer.insert(buffer.end(), bytes, bytes + sizeof(uint16_t));
}

void BinaryWriter::WriteUInt32(uint32_t value)
{
    char bytes[4];
    std::memcpy(bytes, &value, sizeof(uint32_t));
    buffer.insert(buffer.end(), bytes, bytes + sizeof(uint32_t));
}

void BinaryWriter::WriteInt32(int32_t value) {
    char bytes[4];
    std::memcpy(bytes, &value, sizeof(int32_t));
    buffer.insert(buffer.end(), bytes, bytes + sizeof(int32_t));
}

void BinaryWriter::WriteFloat(float value) {
    char bytes[4];
    std::memcpy(bytes, &value, sizeof(float));
    buffer.insert(buffer.end(), bytes, bytes + sizeof(float));
}

void BinaryWriter::WriteString(const std::string& str) {
    uint16_t len = static_cast<uint16_t>(str.length());
    WriteUInt16(len); // 문자열 길이 먼저 기록
    buffer.insert(buffer.end(), str.begin(), str.end());
}

void BinaryWriter::WriteHeader(const PacketHeader& header) {
    WriteUInt16(static_cast<uint16_t>(header.type));
    WriteUInt16(header.size);
}

void BinaryWriter::WriteHeader(EPacketType type, uint16_t size) {
    WriteHeader(PacketHeader(type, size));
}

File: Assets/Script/Packet/BinaryWriter.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace NetworkUtil
{
    public class BinaryWriter
    {
        private List<byte> buffer = new List<byte>();

        public void WriteUInt8(byte value) => buffer.Add(value);

        public void WriteUInt16(ushort value)
        {
            buffer.AddRange(BitConverter.GetBytes(value));
        }

        public void WriteUInt32(uint value)
        {
            buffer.AddRange(BitConverter.GetBytes(value));
        }

        public void WriteInt32(int value)
        {
            buffer.AddRange(BitConverter.GetBytes(value));
        }

        public void WriteFloat(float value)
        {
            buffer.AddRange(BitConverter.GetBytes(value));
        }

        public void WriteString(string str)
        {
            byte[] strBytes = Encoding.UTF8.GetBytes(str);
            ushort len = (ushort)strBytes.Length;
            WriteUInt16(len);
            buffer.AddRange(strBytes);
        }


        public void WriteHeader(EPacketType type, ushort size)
        {
            WriteUInt16((ushort)type);
            WriteUInt16(size);
        }

        public void WriteBytes(byte[] data)
        {
            if (data == null)
                throw new ArgumentNullException(nameof(data));

            buffer.AddRange(data);
        }

        public byte[] ToArray() => buffer.ToArray();
    }
}